### PR TITLE
Add kubernetes discovery

### LINKED ISSUE #139: Add Kubernetes discovery feature
Add a port of JVM Kubernetes discovery. We can use the Kubernetes csharp client to simplify the port, no need to port the JVM HTTP client.

### PR DESCRIPTION
**Pull request by Arkatufus**
Closes #139

### CODE REVIEW

File: src/discovery/kubernetes/Akka.Discovery.KubernetesApi/KubernetesApiServiceDiscovery.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            _log = Logging.GetLogger(system, this);
+            _settings = KubernetesDiscovery.Get(system).Settings;
+            
+            if(_log.IsDebugEnabled)
+                _log.Debug("Settings {0}", _settings);
+        }
+        
+        public override async Task<Resolved> Lookup(Lookup lookup, TimeSpan resolveTimeout)
+        {
+            var labelSelector = _settings.PodLabelSelector(lookup.ServiceName);
```
Comment (Aaronontheweb) [inline src/discovery/kubernetes/Akka.Discovery.KubernetesApi/KubernetesApiServiceDiscovery.cs:49]: Querying pods based on labels - makes sense. Should be easy enough.

File: src/discovery/kubernetes/Akka.Discovery.KubernetesApi/reference.conf [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+######################################################
+
+akka.discovery {
+  # Set the following in your application.conf if you want to use this discovery mechanism:
+  # method = kubernetes-api
+
+  kubernetes-api {
+    class = "Akka.Discovery.KubernetesApi.KubernetesApiServiceDiscovery, Akka.Discovery.KubernetesApi"
+
+    # API server, cert and token information. Currently these are present on K8s versions: 1.6, 1.7, 1.8, and perhaps more
```
Comment (Aaronontheweb) [inline src/discovery/kubernetes/Akka.Discovery.KubernetesApi/reference.conf:12]: We're definitely going to need to do some documentation on this to explain how pods need to be configured in order to allow for them to query the API - it looks to me that by default it _should_ work but there will be situations where it doesn't.

**Review by Aaronontheweb — COMMENTED** [review]
Left some comments

File: src/discovery/kubernetes/Akka.Discovery.KubernetesApi.Tests/multi-container-pod.json [hunk 3]
```diff
@@ -0,0 +1,266 @@
+{
```
[the remark this answers is not in the document]
Comment (Arkatufus) [reply src/discovery/kubernetes/Akka.Discovery.KubernetesApi.Tests/multi-container-pod.json:1]: No, its a dump file of a k8s REST API response

File: src/discovery/kubernetes/Akka.Discovery.KubernetesApi/Extensions.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        public static string GetStringIfDefined(this Configuration.Config config, string key)
+        {
+            var value = config.GetString(key);
+            return string.IsNullOrWhiteSpace(value) || value.Equals($"<{key}>") ? null : value;
+        }
+    }
+
+    internal static class ObjectExtensions
+    {
+        public static T GetOrElse<T>(this T obj, T @default)
```
[the remark this answers is not in the document]
Comment (Arkatufus) [reply src/discovery/kubernetes/Akka.Discovery.KubernetesApi/Extensions.cs:14]: Its just a helper function to make the code less unwieldy to read, it should be inlined in compile time

File: src/discovery/kubernetes/Akka.Discovery.KubernetesApi.Tests/multi-container-pod.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/discovery/kubernetes/Akka.Discovery.KubernetesApi.Tests/multi-container-pod.json:1]: Ah got it

**Comment by Aaronontheweb** [comment]
@Arkatufus merged this - how would you propose adding integration tests?

**Comment by Arkatufus** [comment]
I'm still looking into it, I think I can use the same method used in the k8s C# client repo

**Pull request merged**